docs/decisions/0015-serving-static-assets.rst

```diff
@@ -0,0 +1,125 @@
+
+**Assets must be served from an entirely different domain than the LMS and Studio instances.**
+  To reduce our chance of maliciously uploaded JavaScript compromising LMS and Studio users, user-uploaded assets must live on an entirely different domain from LMS and Studio (i.e. not just another subdomain). So if our LMS is located at ``sandbox.openedx.org``, the files should be accessed at a URL like ``assets.sandbox.openedx.io``.
+
+Operational Requirements
+~~~~~~~~~~~~~~~~~~~~~~~~
+
+**The asset server must be capable of handling high levels of traffic.**
+  Django views are poor choice for streaming files at scale, especially when deploying using WSGI (as Open edX does), since it will tie down a worker process for the entire duration of the response. While a Django-based streaming response is sufficient for small-to-medium sites, we should allow for a more scalable solution that fully takes advantage of modern CDN capabilities.
+
+**Serving assets should not *require* an S3-like object store.**
+  While S3-like object stores are often used in Open edX deployments at scale, our approach shouldn't *require* it. It is acceptable to require an S3-like store for high-scale traffic.
+
+**Serving assets should not *require* ASGI deployment.**
+  Deploying the LMS and Studio using ASGI would likely substantially improve the scalability of a Django-based streaming solution, but migrating and testing this new deployment type for the entire stack is a large task and is considered out of scope for this project.
+
+Decision
+--------
+
+URLs
+~~~~
+
+The format will be: ``https://{asset_server}/assets/apps/{app}/{learning_package_key}/{component_key}/{version}/{filepath}``
+
+The asset server will have a completely different domain from the LMS and Studio, and will not be a subdomain.
+
+A more concrete example: ``https://studio.assets.sandbox.openedx.io/apps/content_libraries/lib:Axim:200/xblock.v1:problem@826eb471-0db2-4943-b343-afa65a6fdeb5/v2/static/images/fig1.png``
+
+The ``version`` can be:
+
+* ``draft`` indicating the latest draft version (viewed by authors in Studio).
+* ``published`` indicating the latest published version (viewed by students in the LMS)
+* ``v{num}`` meaning a specific version–e.g. ``v20`` for version 20.
+
+Site Separation
+~~~~~~~~~~~~~~~
+
+The asset-serving site will run in the same process as the LMS or Studio process the links, so ``studio.assets.sandbox.openedx.io`` would point to the same place that ``studio.sandbox.openedx.org``. Being in the same process will make it easier to implement permissions checks and URL generation code.
+
+The switching between the two would happen with a new Middleware class that is implemented in openedx-learning. We would use this Middleware to make sure that assets are never served through the Studio and LMS URLs, and that no other Studio or LMS views are served through the asset server URL. This will require new configuration settings.
```

Comment:
I'm not sold on the idea of running the asset server in the same process as the LMS and CMS.

First of all, it is neither specific to the LMS nor the CMS so it doesn't make sense to tie it to one, or the other, or both. Second, in production it should preferably be deployed at the edge. By starting with an in-process implementation, we end up having to re-implement everything for the scalable edge version. The edge version can't do in-process permission checks for example.

I think that eventually, the "asset server" should be a really small, lean JavaScript function that uses web standard APIs ([`URL`](https://developer.mozilla.org/en-US/docs/Web/API/URL), `fetch`, `Request`, `Response`), and which makes REST API calls to the LMS as needed to verify authentication and get metadata like content-type etc. This approach can be deployed to almost any modern CDN - Lambda@Edge/CloudFront, CloudFlare Workers, Deno Deploy, Fastly, etc. And those CDNs also each provide a KV store that you can use to cache metadata and permissions.

So for the MVP version, sure, implement it in process in python to keep it simple for now, but maybe put it only in the LMS and make it use a REST API over HTTP to fetch metadata and permissions, so that it will be aligned with the future edge version?

Reply:
I really think Open edX should be deployable without a CDN.

What would you think about implementing this at the nginx layer [using njs](https://github.com/nginx/njs-examples/#fetch) and/or Caddy layer (with a module)? The CDN would still be very useful for edge caching, but asset serving would still be scalable out of the box. Then there's no need for my middleware proposal, and we can stick with the simple permissions+metadata REST endpoint.

I think we would be _requiring_ an object store at that point, because we can't safely expose the storages media files directly on the LMS or Studio domains... unless we want to build in some kind of auth system that only exposes those URLs to the reverse proxy (e.g. token auth).

> First of all, it is neither specific to the LMS nor the CMS so it doesn't make sense to tie it to one, or the other, or both.

I'm not really sure about that. The content libraries will have assets associated with them, and despite the current configuration, the libraries themselves shouldn't be an LMS thing. We might keep a single asset server instead of my proposed subdomains, but I suspect we'll still want to allow for proxying to an endpoint in either Studio for LMS for the permissions check.

Reply:
Yes, something like that would be great, and I think would cover a nice range of dev and small instance setups. I agree CDN shouldn't be a requirement, but I would like to see a fairly scalable default with the ability to add a CDN on top when needed. Maybe I'm overcomplicating it though. It's starting to sound like a lot of layers to develop, deploy, and maintain. And anywhere that we create differences between deployment methodologies (object store vs. local, nginx vs. caddy, CDN vs. not) rather than optional layers that are either present or not, we create potential for bugs and confusion where there's differences in the implementation.

> I suspect we'll still want to allow for proxying to an endpoint in either Studio for LMS for the permissions check.

I guess it could also make sense to disallow access to the draft version from the LMS endpoint, so there could be some differences in how they're configured.

Reply:
> Maybe I'm overcomplicating it though. It's starting to sound like a lot of layers to develop, deploy, and maintain. And anywhere that we create differences between deployment methodologies (object store vs. local, nginx vs. caddy, CDN vs. not) rather than optional layers that are either present or not, we create potential for bugs and confusion where there's differences in the implementation.

So last night I was thinking that maybe we could use a Python service, but a new repo/app that's in an async-based framework that has better streaming performance characteristics than Django. 

But then this morning, I remembered `X-Accel-Redirect`, which I think can cover our needs, and is supported by both nginx and caddy. Instead of a REST response, we'd have Django return the basic content type information for the asset file in the headers, and then set an `X-Accel-Redirect` to the actual file that needs to be done. Then nginx or caddy take things from there with an internal (not browser) redirect.

Reply:
Oooh, that sounds cool!

Reply:
Revised the proposal to use `X-Accel-Redirect`.